helpers/helper_nano.py

```diff
@@ -282,6 +282,216 @@ def ZMuMu_MuSelection(df):
+    df = df.Filter('Sum(isTag)>0')
+
+    df = df.Define('isProbe','abs(Muon_pdgId)==13&&Muon_PassTightId&& (Sum(isTag)>=2||isTag==0)')
+    df = df.Define('dr_mll', 'dR_mll(Muon_pt, Muon_eta, Muon_phi, isTag, isProbe)')
+
+    # Filter on pairs of lepton with DeltaR > 0.5
+    df = df.Filter('''
+    for (unsigned int line = 0; line < dr_mll.size(); line++){
+        for (unsigned int col = 0; col < 2; col++){
+            float DeltaR = dr_mll[line][col][0];
+            if (DeltaR > 0.5){
+                return true;
+            }
+        }
+    }
+    return false;
+    ''')
+
+    df = df.Define('probe_Pt','Muon_pt[isProbe]')
+    df = df.Define('probe_Eta','Muon_eta[isProbe]')
+    df = df.Define('probe_Phi','Muon_phi[isProbe]')
+    df = df.Define('probe_Charge', 'Muon_charge[isProbe]')
+
+    return df
+
+
+def DQMOff_TauSelection(df):
+    '''
+    Selects Z->tautau events passing a single muon trigger. Defines probe tau pt/eta/phi
+    The selections are made to match with DQM Off selections
+    '''
+    df = df.Filter('HLT_IsoMu20')
+
+    # Trigged on a Muon (probably redundant)
+    df = df.Filter('''
+    bool trigged_on_mu = false;
+    for (unsigned int i = 0; i < TrigObj_id.size(); i++){
+        if(TrigObj_id[i] == 13) trigged_on_mu = true;
+    }
+    return trigged_on_mu;
+    ''')
+
+    # Charge 
+    # L1Mu_hwCharge = 0 corresponds to Muon_charge = +1
+    # L1Mu_hwCharge = 1 corresponds to Muon_charge = -1
+    # df = df.Define('L1Mu_charge', 'charge_conversion(L1Mu_hwCharge)')
+
+    # --------------------------------------------------------
+    # Match L1Mu to trig obj with id == 13 -> L1Mu_trig_idx
+    # with L1Mu_trig_idx -> check filter bits -> L1Mu_passHLT
+    # --------------------------------------------------------
+
+    # TrigObj matching
+    df = df.Define('Muon_trig_idx', 'MatchObjToTrig_DQMOff(Muon_eta, Muon_phi, TrigObj_pt, TrigObj_eta, TrigObj_phi, TrigObj_id, 13, TrigObj_filterBits, 3, 0.6)')
+    df = df.Define('Muon_passHLT_IsoMu24', 'trig_is_filterbit_set_DQMOff(Muon_trig_idx, TrigObj_filterBits, 3)')
+
+    df = df.Define('Muon_PassTightId','Muon_pfIsoId>=3&&Muon_mediumPromptId') 
+    df = df.Define('Muon_MassId', 'pass_muon_met_mass_below30(Muon_pt, Muon_phi, MET_pt, MET_phi)')
+
+    df = df.Define('isTag','Muon_pt>24&&abs(Muon_eta)<2.1&&abs(Muon_pdgId)==13&&Muon_PassTightId&&Muon_passHLT_IsoMu24&&Muon_MassId')
+    df = df.Filter('Sum(isTag)>0')
+
+    df = df.Define('tagIdx', 'first_tagmuon_idx(isTag)')
+    df = df.Filter('tagIdx!=-1')
+
+    df = df.Define('tagPt', 'Muon_pt[tagIdx]')
+    df = df.Define('tagEta', 'Muon_eta[tagIdx]')
+    df = df.Define('tagPhi', 'Muon_phi[tagIdx]')
+    df = df.Define('tagMass', 'Muon_mass[tagIdx]')
+    df = df.Define('tagCharge', 'Muon_charge[tagIdx]')
+
+    nEvts = df.Count().GetValue()
+    print("There are {} events after finding tag muon".format(nEvts))
+
```

Comment:
These two lines are triggering an event loop here. They should come later.

Reply:
These are just for debugging/testing purpose. Have been removed.